graph2tac/loader/predict_server.py

```diff
@@ -37,6 +37,44 @@ def apply_temperature(confidences: ArrayLike, temperature: float) -> NDArray[np.
     res /= res.sum() + (1-confidences.sum())**(1/temperature)
     return res
 
+class ResponseHistory:
+    """Records response history for testing."""
+    data: dict
+    """JSON compatible dictionary which stores all the history."""
+    def __init__(self, recording_on: bool):
+        """
+        :param recording_on: Record results if True, otherwise do nothing.
+        """
+        self._recording_on = recording_on
```

Comment:
It is strange to me to have a recording class that might not be recording. Related to the question below regarding the wrapper for `send`. I would do the logic for recording/not-recording in such a wrapper. (But whatever, I don't care enough about this for an extended discussion.)

Reply:
I am rather unsure if we need two different mechanisms for recording -- `response_history`, and `record_file`. Could we keep only one of them?

Reply:
Mhm, Jason suggested this as well. But if we do this, then we have to generate a new `trace.bin` file whenever something changes in the servers responses. This is possible, I guess. But then there still needs to be logic somewhere that compares an old trace to a new trace.

This is all possible, but it's not clear to me what the API for this should look like. Do you have a suggestion?

Reply:
@LasseBlaauwbroek I'll see what I can do.

Reply:
> A note about using yaml for test output: Yaml is a notoriously complicated format. (Yes, once parsed it is a superset of json, but the printed format is ridiculously complicated.) See [the yaml document from hell](https://ruudvanasseldonk.com/2023/01/11/the-yaml-document-from-hell). It might be kind of dangerous to use this for tests.
> 
> (I would personally probably never use yaml for anything unless it was already part of the system. Instead I would look at toml or something else.)

That is a good point about yaml.  Maybe in hindsight we shouldn't have invested in yaml for parameters (and went with toml).  But for this particular use case, my main reason is that it makes the test result files readible and diffable.  The later is actually really useful I'm finding.  If we have to change one of these, it is really easy to use git diff to see what we actually changed.  We can't do that with the default printing of JSON, but I guess we could switch to another JSON output format where each piece of data is on its own line.  I just liked YAML since it does it automatically.

And to be clear, the tests are not comparing the files themselves.  They are importing the files into dictionaries and comparing those using pytest's ability to compare dictionaries (and it's ability to compare floats up to a tolerance).

Reply:
Yes, readable output is a good feature! But in hindsight, I would have used toml, or something else. Whatever, either keep it or switch to toml, up to you. I just thought you should be aware of the pitfalls.

Reply:
@mirefek The reason for having two recording methods now is as follows.  I originally also just wanted one, but Lasse convinced me this is better and I agree.  Here are what the two recording mechanisms do:
* `--record` with record both incoming and outgoing messages when interacting with coq via stdin or tcp.  It is stored in a binary file.  If I want to make a new recording, I need to have coq running (or fake coq) at least, so I need an external processes running at the same time.  This makes it a bit hard to use in tests where I want to be running our server directly from python and also to be able to update the fingerprints if we change the network.
* `--replay` isn't technically about recording, but about replaying the above record file.  So `--replay` is an alternative to `--tcp` and the default stdin communication channels.  I just currently have it set up to also record outgoing messages only when `--replay` is on to avoid a memory leak were we record all outgoing messages in a long run.  The recording that is happening here is for the tests (although we could use it for other debugging reasons).  I only record the outgoing messages, and I record what I need to know for the tests.  I'm recording to a dictionary which is returned by `main` and can be used in the test file (just like I do for the hmodel and tfgnn train.  It is easy to then do the tests with this dictionary, comparing it to what was previously stored (and comparing the floats up to an error tolerance).  I was hoping I could just do something like `msg.to_dict()` to get the dictionaries for each message, but this is not the case, so it does require a bit more work. Also, I don't have to interact with coq or fakecoq at all here and it is easy to overwrite the tests whenever the network changes from inside `pytest`.

If we had only one system, it would probably look like follows:
* We update the mechanism I use for `--replay` to also allow `--record` to be usable at the same time.  So we read in stored incoming messages from one binary file, and store the new outgoing messages into another.  Lasse could add this functionality if we want.
* Then when we run a test, we don't return the history as a dictionary, but save it to file in the temp directory and the tests read that in.
* The tests would have to read in that record file and convert it to a dictionary for comparison.  So in this sense it would require again writing all that code to convert messages to dictionaries and we wouldn't save much complexity.
* If the tests fail and we use --overwrite, then we replace the old binary record file with this one.  We wouldn't be able to easily see how the test results have changed.

Personally, I like the current approach, and agree with Lasse's original suggestion.

Reply:
> We update the mechanism I use for --replay to also allow --record to be usable at the same time.

As per a comment of mine earlier in this thread: I just did this.

Reply:
[One more change](https://github.com/coq-tactician/coq-tactician-reinforce/commit/623952029b0b5e997c08634ba09bb549a0015a0a) to the API, to support @mirefek's preference for a 'unified approach': You can now compare the recorded response with the actual server response through a Callable:
```python
def capnp_message_generator_from_file_lowlevel(
        message_file: BinaryIO,
        check : Callable[[Any, Any, Any], None] | None = None) -> (
        Generator[pytact.graph_api_capnp_cython.PredictionProtocol_Request_Reader,
                  capnp.lib.capnp._DynamicStructBuilder, None]):
```
The callable of `check` will be called like
```python
check(request_message, server_response_message, recorded_response_message)
```
I believe that this is enough to implement the 'unified approach'. Wether or not to actually do that, I'll leave for @mirefek and @jasonrute to discuss.

Reply:
Thanks! I would rather not spend time on it with this commit, and get back to it in a later refactor of `predict_server` (say when I would get to work on nested contexts).